botocore/utils.py

```diff
@@ -644,24 +646,32 @@ def fix_s3_host(request, signature_version, region_name, **kwargs):
         return
     elif not _allowed_region(region_name):
         return
+
+    # We need to respect custom endpoint urls, but unfortunately we have no
+    # way at this point to know how the url got set. We unfortunately don't have
+    # access to the endpoint resolver here, so we can't do a data-driven test.
+    # A regular expression is therefore used to determine if the request url is
+    # an S3 url.
+    endpoint_url = None
+    match = S3_ENDPOINT_RE.match(request.url)
```

Comment:
We could spin up an endpoint resolver, but that would require a load from disk...